src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseExportService.java

```diff
@@ -149,6 +157,116 @@ private void deleteTempZipRepoFiles(List<Path> pathsToZipeedRepos) {
         }
     }
 
+    /**
+     * downloads all repos of the exercise and runs JPlag
+     *
+     * @param programmingExerciseId the id of the programming exercises which should be checked
+     * @return a zip file that can be returned to the client
+     * @throws ExitException is thrown if JPlag exits unexpectedly
+     * @throws IOException is thrown for file handling errors
+     */
+    public File checkPlagiarism(long programmingExerciseId) throws ExitException, IOException {
+        ProgrammingExercise programmingExercise = programmingExerciseRepository.findWithAllParticipationsById(programmingExerciseId).get();
+
+        programmingExercise.getStudentParticipations().parallelStream().forEach(participation -> {
+            var programmingExerciseParticipation = (ProgrammingExerciseParticipation) participation;
+            try {
+                if (programmingExerciseParticipation.getRepositoryUrlAsUrl() == null) {
+                    log.warn("Ignore participation " + participation.getId() + " for export, because its repository URL is null");
+                    return;
+                }
+                Repository repo = gitService.getOrCheckoutRepository(programmingExerciseParticipation, REPO_DOWNLOAD_CLONE_PATH);
+                gitService.resetToOriginMaster(repo); // start with clean state
```

Comment:
If I run this while an exercise is still open: Would this discard changes in the online code editor?

Reply:
no, this is done in a separate path REPO_DOWNLOAD_CLONE_PATH